### DISCUSSION TITLE
Replace Set object with Map object

### DISCUSSION THREAD

**Discussion by mateonunez** [oramasearch/orama#34] [Ideas]
I was working on #12 and #13. I stashed my code a lot of times looking to keep the same architecture of inserted documents. So hard or stressful for Lyra.

I created a new architecture for inserted documents, replacing the `Set` with a `Map<string, number>` (where the number could be replaced with a typed object that will contain _metadata_).

![image](https://user-images.githubusercontent.com/11861080/179832812-75af5ab3-a234-4c80-84bf-3cba70bb255c.png)

![image](https://user-images.githubusercontent.com/11861080/179833017-925896ed-940c-4d2f-af5d-40d8cc7232fd.png)

The tests are awesome:

**main branch:**

![image](https://user-images.githubusercontent.com/11861080/179834042-26fbfa72-a16a-42b0-b14d-59b262c159dc.png)

**with new implementations:**

![image](https://user-images.githubusercontent.com/11861080/179834178-cbb0ffbf-31cb-4236-ad52-17629cd62fd5.png)

> The elapsed time was computed before the insert and after the search methods. I've inserted 1000 documents.

This could be an idea to implement more _metadata_ when a document is indexed. 

I also created a [branch](https://github.com/nearform/lyra/compare/main...mateonunez:lyra:feat/occurrences-sort?diff=unified) (in WIP and breaking in some cases) to show how could it work.

**Comment by mateonunez** [comment]
I've fixed some tests in the branch (still in WIP). I also did some benchmarks.

**feat/occurrences-sort benchmark** (_replaced 'cause checkout on the wrong branch_):
[https://pastebin.com/raw/90V6r1ZL](https://pastebin.com/raw/90V6r1ZL)

**main benchmark**
[https://pastebin.com/raw/u4ARt3us](https://pastebin.com/raw/u4ARt3us)

I think this implementation can work properly, it doesn't have a big impact on indexing or searching performance, and also could bring an extension toward the functionality that will be able to be performed by Lyra. For example, we can implement a `metadata` object. I also created a [branch](https://github.com/nearform/lyra/compare/main...mateonunez:lyra:feat/occurrences-sort-with-metadata) to show you how it could work.

**Comment by micheleriva** [comment]
@mateonunez as wrote in https://github.com/nearform/lyra/discussions/56, we had to remove all the `Map` objects to perform `dpack` serialization on the indexes. As for now, it is not possible to add more `Map` objects to the codebase 😞
